electroncash_gui/qt/utils/validators.py

```diff
@@ -54,15 +70,64 @@ def validate(self, inputStr: str, pos: int) -> QValidator.State:
     def setRedBorder(self, state):
         parent = self.parent()
         if isinstance(parent, QLineEdit):
-            if state == QValidator.Acceptable:
-                parent.setStyleSheet('')
-            else:
-                parent.setStyleSheet('QLineEdit { border: 1px solid red }')
+            mark_widget(state, parent)
+
 
 class UserPortValidator(PortValidator):
     """
     Checks that a given port is either a high port (from 1024 to 65535) or zero.
     Additionally provides a callback for when the validation state changes.
     """
+
     def __init__(self, parent, accept_zero=False):
         super().__init__(parent, 1024, accept_zero)
+
+
+class HostValidator(QValidator):
+    """
+    Validates a host string, accepts either IPV4, IPV6 or domain names
+    """
+
+    stateChanged = pyqtSignal(QValidator.State)
+
+    def __init__(self, parent):
+        super().__init__(parent)
+        if not isinstance(parent, QLineEdit):
+            warnings.warn(RuntimeWarning("HostValidator must be passed a QLineEdit parent"))
+        self.stateChanged.connect(self.setRedBorder)
+
+    _domain_pattern = re.compile(r"^[a-zA-Z0-9.-]+$")
+
+    def validate(self, inputStr: str, pos: int) -> typing.Tuple[QValidator.State, str, int]:
+        def set_state(state):
+            self.stateChanged.emit(state)
+            return state, inputStr, pos
+
+        if len(inputStr) == 0:
+            return set_state(QValidator.Intermediate)
+
+        # Check for valid IP address
+        try:
+            ipaddress.ip_address(inputStr)
+            return set_state(QValidator.Acceptable)
+        except ValueError:
+            pass
+
+        # We try regex first to save a DNS lookup on clearly invalid names
+        if not self._domain_pattern.match(inputStr):
+            return set_state(QValidator.Invalid)
+
+        # Check for resolvable DNS
```

Comment:
I suspect this is where the slow is.

Any chance we can just not do this?  I mean .. isn't it fine to just validate IP and something that *could* be a hostname and call it a day?

Reply:
Yea, I'd be fine with that too. Will adjust soon.

Reply:
I'll also adjust the regex to be a bit more correct.

Reply:
Updated

Reply:
thanks man